File: backend/app/core/upgrades/effects/regeneration.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Regeneration(UpgradeImplementation):
    id = 'regeneration'
    name = 'Auto Repair'
    description = 'Regenerate 1 HP/sec'
    rarity = 'epic'
    type = 'effect'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.effects.register_effect('regen', {'amount': 1})

    def validate(self, recorded_stats: dict) -> bool:
        health_recovered = recorded_stats.get('health_recovered', 0)
        return health_recovered > 5

File: frontend/src/game/upgrades/effects/regeneration.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { GameManager } from '../../core/GameManager'
import type { Player } from '../../entities/Player'
import { RarityID, UpgradeTypeID } from '../../data/ID'

export const RegenerationDef: UpgradeDef = {
  id: "regeneration",
  name: "Auto Repair",
  description: "Regenerate 1 HP/sec",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.Effect,
  cost: 20,
  effect: "regen",
  effectValue: 1,
  stackable: true,
  maxStacks: 3,
}

export class Regeneration extends Upgrade {
  onApply(): void {}

  updatePlayer(_player: Player, delta: number): void {
    GameManager.heal(this.def.effectValue! * (delta / 1000))
  }
}
